### PR TITLE
[bug-fix] Change entropy computation and loss reporting in Torch to match TF

### PR DESCRIPTION
**Pull request by ervteng**
### Proposed change(s)

Entropy was being treated differently in Torch than in TF (TF took the sum of all the entropy across the actions, and Torch did the mean). Furthermore, TF took the absolute value of policy loss in PPO. This PR makes both of these match TF for easier debugging, consistent entropy bonus magnitude, and more consistent user behavior. 

For continuous actions in PPO, entropy was also a single value - and not reported for every action in the batch. This PR also fixes that bug. 

### Types of change(s)

- [x] Bug fix
- [ ] New feature
- [ ] Code refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [ ] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/master/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [ ] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/master/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md) (if applicable)

### CODE REVIEW

**Review by andrewcoh — COMMENTED** [review]
Does this line make sense any more https://github.com/Unity-Technologies/ml-agents/blob/master/ml-agents/mlagents/trainers/ppo/optimizer_torch.py#L180?

File: ml-agents/mlagents/trainers/ppo/optimizer_torch.py [hunk 1]
```diff
@@ -187,7 +187,7 @@ def update(self, batch: AgentBuffer, num_sequences: int) -> Dict[str, float]:
 
         self.optimizer.step()
         update_stats = {
-            "Losses/Policy Loss": policy_loss.item(),
+            "Losses/Policy Loss": torch.abs(policy_loss).item(),
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/ppo/optimizer_torch.py:190]: Why is this needed ?

**Comment by vincentpierre** [comment]
> Does this line make sense any more https://github.com/Unity-Technologies/ml-agents/blob/master/ml-agents/mlagents/trainers/ppo/optimizer_torch.py#L180?

What do you mean? This is the entropy regularization, I think we need it. Are you referring to the use of `masked_mean` ?

File: ml-agents/mlagents/trainers/ppo/optimizer_torch.py (hunk 1 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/ppo/optimizer_torch.py:190]: > Does this line make sense any more https://github.com/Unity-Technologies/ml-agents/blob/master/ml-agents/mlagents/trainers/ppo/optimizer_torch.py#L180?

The masked_mean was actually _not_ being used properly before (it was just a single value). It is now, as the entropy is BxN.

**Comment by ervteng** [comment]
> Does this line make sense any more https://github.com/Unity-Technologies/ml-agents/blob/master/ml-agents/mlagents/trainers/ppo/optimizer_torch.py#L180?

The masked_mean was actually _not_ being used properly before (it was just a single value). It is now, as the entropy is BxN.

File: ml-agents/mlagents/trainers/ppo/optimizer_torch.py (hunk 1 above)
Comment (vincentpierre) [reply ml-agents/mlagents/trainers/ppo/optimizer_torch.py:190]: I do not see why that helps much. The loss is roughly `mean(Prob x Advantage)`. This loss can fluctuate above/bellow zero. What we report is then `mean(abs(mean(Prob x Advantage)))`  (Because we do the mean of the aggregated stats) which is just very weird. Maybe we can report : 
 - `mean(mean(abs(Prob x Advantage)))` (Not sure if it makes sense)
 - `mean(mean(Prob x Advantage))` (remove the abs, is what it says it is)
 - `mean(std_dev(Prob x Advantage))` If the fluctuations are large, the std_dev should tell us
 - `mean(max(abs(Prob x Advantage)))` If we want to know what is the largest error we have

**Comment by andrewcoh** [comment]
> > Does this line make sense any more https://github.com/Unity-Technologies/ml-agents/blob/master/ml-agents/mlagents/trainers/ppo/optimizer_torch.py#L180?
> 
> What do you mean? This is the entropy regularization, I think we need it. Are you referring to the use of `masked_mean` ?

To the use of masked_mean.

File: ml-agents/mlagents/trainers/ppo/optimizer_torch.py (hunk 1 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/ppo/optimizer_torch.py:190]: TBH none of these are really interpretable from the perspective of what we'd normally expect from "loss" (i.e. loss goes down over time). Seems like the last one might get the right behavior, but it's a real stretch to call it a "Policy Loss". Maybe we can go that route and call it "Policy Error" or something similar. 

Otherwise, just going with `mean(mean(Prob x Advantage))` is the most correct, but maybe not useful.

Comment (ervteng) [reply ml-agents/mlagents/trainers/ppo/optimizer_torch.py:190]: I've been thinking more about this - I think we should leave it with the `abs` at least until the first release after PyTorch default switch, so that users who are comparing TF and Torch will have the exact same plots with no surprises.

Comment (vincentpierre) [reply ml-agents/mlagents/trainers/ppo/optimizer_torch.py:192]: Ok, but we should revisit once TF is deprecated.

Comment (ervteng) [reply ml-agents/mlagents/trainers/ppo/optimizer_torch.py:192]: 100% for sure. I think we probably want both the `std_dev` and `mean` at the end of the day.

**Review by vincentpierre — APPROVED** [review]
